lib/spack/spack/mixins.py

```diff
@@ -0,0 +1,208 @@
+        phases_to_be_copied = list(
+            PackageMixinsMeta._add_method_before.keys()
+        )
+        phases_to_be_copied += list(
+            PackageMixinsMeta._add_method_after.keys()
+        )
+
+        for phase in phases_to_be_copied:
+
+            attr_name = attr_fmt.format(phase)
+
+            # Here we want to get the attribute directly from the class (not
+            # from the instance), so that we can modify it and add the mixin
+            # method to the pipeline.
+            phase = getattr(cls, attr_name)
+
+            # Due to MRO, we may have taken a method from a parent class
+            # and modifying it may influence other packages in unwanted
+            # manners. Solve the problem by copying the phase into the most
+            # derived class.
+            setattr(cls, attr_name, phase.copy())
+
+        # Insert the methods in the appropriate position
+        # in the installation pipeline.
+
+        for phase in PackageMixinsMeta._add_method_before:
+
+            attr_name = attr_fmt.format(phase)
+            phase_obj = getattr(cls, attr_name)
+            fn_list = PackageMixinsMeta._add_method_after[phase]
+
+            for f in fn_list:
+                phase_obj.run_before.append(f)
+
+        # Flush the dictionary for the next class
+        PackageMixinsMeta._add_method_before.clear()
+
+        for phase in PackageMixinsMeta._add_method_after:
+
+            attr_name = attr_fmt.format(phase)
+            phase_obj = getattr(cls, attr_name)
+            fn_list = PackageMixinsMeta._add_method_after[phase]
+
+            for f in fn_list:
+                phase_obj.run_after.append(f)
+
+        # Flush the dictionary for the next class
+        PackageMixinsMeta._add_method_after.clear()
+
+        super(PackageMixinsMeta, cls).__init__(name, bases, attr_dict)
+
+
+def filter_compiler_wrappers(*files, **kwargs):
+    """Substitutes any path referring to a Spack compiler wrapper with the
+    path of the underlying compiler that has been used.
+
+    If this isn't done, the files will have CC, CXX, F77, and FC set to
+    Spack's generic cc, c++, f77, and f90. We want them to be bound to
+    whatever compiler they were built with.
+
+    Args:
+        *files: files to be filtered relative to the search root (which is,
+            by default, the installation prefix)
+
+        **kwargs: allowed keyword arguments
+
+            after
+                specifies after which phase the files should be
+                filtered (defaults to 'install')
+
+            relative_root
+                path relative to prefix where to start searching for
+                the files to be filtered. If not set the install prefix
+                wil be used as the search root. **It is highly recommended
+                to set this, as searching from the installation prefix may
+                affect performance severely in some cases**.
```

Comment:
@tgamblin I started having a look at this, and have some questions.

> I think it's unintuitive that this would search by default and we have to add some docs to relative_root to "strongly recommend setting this" because of it.

If you refer to the lines above when you say  "strongly recommend setting this" they don't apply anymore, as we don't recurse by default. I think I need to remove the warning from the docstring (it was added when the default was to search recursively into subdirectories). 

> instead the default is just to specify relative paths, with no searching.

If we remove the use of `find` by default we'll also remove globbing - which was a previous request :smile: 

----

To summarize:

- right now we do a non-recursive globbing search by default, that starts at the `prefix` of the spec
- setting `relative_root` makes the search start from another location, relative to the prefix
- not setting `relative_root` could be a problem only for people that **explicitly** set a recursive search 

I would be for just renaming `relative_root` to `search_from` and update the docstring above, but I am not sure if I  am overlooking some of your concerns?

Reply:
@tgamblin ping